masonry/src/doc/color_rectangle.rs

```diff
@@ -0,0 +1,312 @@
+// ---
+use masonry::core::{BoxConstraints, LayoutCtx};
+// ---
+use masonry::accesskit::{Node, Role};
+use masonry::core::{AccessCtx, PaintCtx, PropertiesRef};
+use masonry::kurbo::Affine;
+use masonry::peniko::Fill;
+use masonry::vello::Scene;
+// ---
+use masonry::core::WidgetId;
+use tracing::{Span, trace_span};
+// ---
+use masonry::core::{ChildrenIds, RegisterCtx};
+// ---
+use masonry::core::WidgetMut;
+// ---
+use masonry::properties::Background;
+
+// ---
+
+pub struct ColorRectangle {
+    size: Size,
+    color: Color,
+}
+
+impl ColorRectangle {
+    pub fn new(size: Size, color: Color) -> Self {
+        Self { size, color }
+    }
+}
+
+// ---
+
+impl ColorRectangle {
+    pub fn set_color(this: &mut WidgetMut<'_, Self>, color: Color) {
+        this.widget.color = color;
+        this.ctx.request_paint_only();
+    }
+
+    pub fn set_size(this: &mut WidgetMut<'_, Self>, size: Size) {
+        this.widget.size = size;
+        this.ctx.request_layout();
+    }
+}
+
+// ---
+
+#[derive(Debug)]
+struct ColorRectanglePress;
+
+impl Widget for ColorRectangle {
+    fn on_pointer_event(
+        &mut self,
+        ctx: &mut EventCtx<'_>,
+        _props: &mut PropertiesMut<'_>,
+        event: &PointerEvent,
+    ) {
+        match event {
+            PointerEvent::Down {
+                button: Some(PointerButton::Primary),
+                ..
+            } => {
+                ctx.submit_action(ColorRectanglePress);
+            }
+            _ => {}
+        }
+    }
+
+    fn on_text_event(
+        &mut self,
+        _ctx: &mut EventCtx<'_>,
+        _props: &mut PropertiesMut<'_>,
+        _event: &TextEvent,
+    ) {
+    }
+
+    fn on_access_event(
+        &mut self,
+        ctx: &mut EventCtx<'_>,
+        _props: &mut PropertiesMut<'_>,
+        event: &AccessEvent,
+    ) {
+        match event.action {
+            accesskit::Action::Click => {
+                ctx.submit_action(ColorRectanglePress);
+            }
+            _ => {}
+        }
+    }
+
+    // ---
+
+    fn on_anim_frame(
+        &mut self,
+        _ctx: &mut UpdateCtx<'_>,
+        _props: &mut PropertiesMut<'_>,
+        _interval: u64,
+    ) {
+    }
+    fn update(
+        &mut self,
+        _ctx: &mut UpdateCtx<'_>,
+        _props: &mut PropertiesMut<'_>,
+        _event: &Update,
+    ) {
+    }
+
+    // ---
+
+    fn layout(
+        &mut self,
+        _ctx: &mut LayoutCtx<'_>,
+        _props: &mut PropertiesMut<'_>,
+        bc: &BoxConstraints,
+    ) -> Size {
+        bc.constrain(self.size)
+    }
+
+    // ---
+
+    #[cfg(false)] // We show two `paint` implementations; check that both parse.
+    fn paint(&mut self, ctx: &mut PaintCtx<'_>, _props: &PropertiesRef<'_>, scene: &mut Scene) {
+        let rect = ctx.size().to_rect();
+        scene.fill(
+            Fill::NonZero,
+            Affine::IDENTITY,
+            self.color,
+            Some(Affine::IDENTITY),
+            &rect,
+        );
+    }
```

Comment:
I think that the pattern around line 210 might be worth repeating here, so that the impl compiling is checked, not just parsing

Reply:
I get your point, but also part of the pattern I'm following is to make it easy to update the tutorial examples by just copy-pasting from the file. Moving the `paint` code into a separate block would break that. I guess I could move the other paint block though.

Reply:
I think it would make sense for them both to use the same pattern, whether that's neither being checked or both being checked.
But whatever you think works; this is a non-blocking concern (where the `doc(hidden)` question was/is blocking). 

Alternatively, we could make this whole block not be an `impl Widget` block, but that risks missing something if new widget methods are added.